replaces `unsafe.Pointer(&slice[0])` with `unsafe.SliceData(slice)`

ibeckermayer opened the pull request:
Go 1.20 added [`unsafe.SliceData`](https://pkg.go.dev/unsafe#SliceData) which provides a cleaner API for this type of operation.

Commit:
replaces unsafe.Pointer(&slice[0]) with unsafe.SliceData(slice)

fspmarshall reviewed (approved):
From `SliceData` docs:
> - If cap(slice) > 0, SliceData returns &slice[:1][0].
> - If slice == nil, SliceData returns nil.
> - Otherwise, SliceData returns a non-nil pointer to an unspecified memory address.

That third bullet point is concerning.  Prior to this PR we would panic on a zero-capacity (or nil) slice, which ultimately seems saner than whatever "unspecified memory address" means.

I don't see anywhere that we'd obviously hit this case currently, but I think its worth adding checks so that in the future if some issue is introduced that would hit this case, we instead ensure that we are generating a panic or error instead.

LGMT otherwise.

Commit:
Adds utils.UnsafeSliceData
which converts the unsafe.SliceData edge case which returns a sketchy
"a non-nil pointer to an unspecified memory address" to an error.

ibeckermayer commented:
> That third bullet point is concerning. Prior to this PR we would panic on a zero-capacity (or nil) slice, which ultimately seems saner than whatever "unspecified memory address" means.

Excellent point with the caveat that in certain cases like [here](https://github.com/gravitational/teleport/pull/28625/files#diff-27da86cebdfa8870a5999b5cb051103a7cfc23e41131139feab9b26e935a5915L544-L547) we handled the `nil` case without panicking.

See 6e6af094b85729cebd5e8a3a2f4b420734496a60 -- I created a util function that returns an error before letting us mistakenly do anything with an "unspecified memory address".
